### ISSUE TITLE
Confusing behavior and documentation of tabular test dataloader

### ISSUE THREAD

**Issue by byte01** [fastai/fastai#2911]
Feature requests should first be proposed on [the forum](https://forums.fast.ai). 

**Link to forum discussion.**

https://forums.fast.ai/t/kaggle-titanic-and-decisiontreeregressor/79655/5?u=maxim

**Is your feature request related to a problem? Please describe.**
Currently tst_dl = dls.valid.new(to_tst) doesn't work when a test set has some missing values other than the train set.

**Describe the solution you'd like**
~~A thinking machine, capable of AGI.~~
Please provide a clear documented way to predict on a test set with different missing values than training set with tabular data loader. Maybe there is a way to automatically add proper NaN columns to test and train sets, without having to do it manually?

**Describe alternatives you've considered**
Currently

dls = to.dataloaders()
test_dl = dls.test_dl(df_test)
m.predict(test_dl.dataset.xs[to_keep].drop(to_drop, axis=1))

works (on kaggle House Prices: Advanced Regression Techniques) if I add a fake row with NaNs (in places where NaNs appear in train+test dataset) to my train set.

While
to_test = to.valid.new(df_test)
m.predict(to_test.xs[to_keep].drop(to_drop, axis=1))

doesn't work, but in documentation it is shown as one of the ways to get test predictions:
https://docs.fast.ai/tabular.core

**Additional context**
The way I add a fake row with NaNs:

df_all = df.append(df_test)
display_all(df_all.isnull().sum().sort_index()/len(df_all))

dep_var = ‘SalePrice’

nas = df_all.isnull().sum()
nas[dep_var] = 0
nas

df_mode = df.mode(axis=0,dropna=False)
df_mode

fake_nan_row = np.where(nas == 0, df_mode, np.NaN)
pd.DataFrame(fake_nan_row)

df = df.append(pd.DataFrame(fake_nan_row,columns=df_mode.columns,index=[len(df)]))
display_all(df)

fake_idx = len(df)-1
fake_idx

def idx_move_train(splits, idx):
    if idx in splits[0]:
        return splits
    else:
        return (splits[0]+[idx], splits[1].remove(idx))
splits =  RandomSplitter(valid_pct=0.2)(range_of(df))

splits = idx_move_train(splits, fake_idx)

**Comment by jph00** [comment]
@muellerzr would you be open to taking a look at this? Do you already have a notebook that might be suitable?

**Comment by muellerzr** [comment]
@jph00 I don't think the solution to this is more than simply a note saying that it will do so, as adding "blank" NaN's would lead to bad results. If you agree with this thinking I can be a bit more verbose

**Comment by jph00** [comment]
I'd be most happy to see your more verbose thinking :)

**Comment by muellerzr** [comment]
This should be able to close. I've posted in the documentation in both tabular.core as well as tabular.tutorial this explaination: 

> Note: Since machine learning models can't magically understand categories it was never trained on, the data should reflect this. If there are different missing values in your test data you should address this before training

**Comment by byte01** [comment]
Thank you!
